### DISCUSSION TITLE
Deleting users from organisation using directory connector not working

### DISCUSSION THREAD

**Discussion by bokkabonga** [dani-garcia/vaultwarden#3689] [Q&A]
### Subject of the issue

Deleting users from organisation using directory connector not working

### Your environment (Generated via diagnostics page)
* Vaultwarden version: v1.29.0-61f90818
* Web-vault version: v2023.5.0
* OS/Arch: linux/x86_64
* Running within Docker: true (Base: Debian)
* Environment settings overridden: true
* Uses a reverse proxy: true
* IP Header check: true (X-Real-IP)
* Internet access: true
* Internet access via a proxy: false
* DNS Check: true
* Browser/Server Time Check: true
* Server/NTP Time Check: true
* Domain Configuration Check: true
* HTTPS Check: true
* Database type: MySQL
* Database version: 10.4.20-MariaDB-1:10.4.20+maria~focal
* Clients used: 
* Reverse proxy and version: 
* Other relevant information: 

### Config (Generated via diagnostics page)
<details><summary>Show Running Config</summary>

**Environment settings which are overridden:** INVITATION_ORG_NAME, DISABLE_2FA_REMEMBER, AUTHENTICATOR_DISABLE_TIME_DRIFT, LOG_TIMESTAMP_FORMAT, EMAIL_EXPIRATION_TIME, EMAIL_ATTEMPTS_LIMIT

```json
{
  "_duo_akey": null,
  "_enable_duo": false,
  "_enable_email_2fa": false,
  "_enable_smtp": true,
  "_enable_yubico": true,
  "_icon_service_csp": "",
  "_icon_service_url": "",
  "_ip_header_enabled": true,
  "_smtp_img_src": "cid:",
  "admin_ratelimit_max_burst": 3,
  "admin_ratelimit_seconds": 300,
  "admin_session_lifetime": 20,
  "admin_token": "***",
  "allowed_iframe_ancestors": "",
  "attachments_folder": "data/attachments",
  "authenticator_disable_time_drift": false,
  "data_folder": "data",
  "database_conn_init": "",
  "database_max_conns": 10,
  "database_timeout": 30,
  "database_url": "*****://***************************************************",
  "db_connection_retries": 10,
  "disable_2fa_remember": true,
  "disable_admin_token": false,
  "disable_icon_download": false,
  "domain": "*****://****************",
  "domain_origin": "*****://****************",
  "domain_path": "",
  "domain_set": true,
  "duo_host": null,
  "duo_ikey": null,
  "duo_skey": null,
  "email_attempts_limit": 3,
  "email_expiration_time": 600,
  "email_token_size": 6,
  "emergency_access_allowed": true,
  "emergency_notification_reminder_schedule": "0 3 * * * *",
  "emergency_request_timeout_schedule": "0 7 * * * *",
  "enable_db_wal": false,
  "event_cleanup_schedule": "0 10 0 * * *",
  "events_days_retain": null,
  "extended_logging": true,
  "helo_name": null,
  "hibp_api_key": "***",
  "icon_blacklist_non_global_ips": true,
  "icon_blacklist_regex": null,
  "icon_cache_folder": "data/icon_cache",
  "icon_cache_negttl": 259200,
  "icon_cache_ttl": 2592000,
  "icon_download_timeout": 10,
  "icon_redirect_code": 302,
  "icon_service": "internal",
  "incomplete_2fa_schedule": "30 * * * * *",
  "incomplete_2fa_time_limit": 3,
  "invitation_expiration_hours": 120,
  "invitation_org_name": "Bitwarden",
  "invitations_allowed": true,
  "ip_header": "X-Real-IP",
  "job_poll_interval_ms": 30000,
  "log_file": "/data/bitwarden.log",
  "log_level": "info",
  "log_timestamp_format": "%Y-%m-%d %H:%M:%S.%3f",
  "login_ratelimit_max_burst": 10,
  "login_ratelimit_seconds": 60,
  "org_attachment_limit": 20480,
  "org_creation_users": "***",
  "org_events_enabled": true,
  "org_groups_enabled": true,
  "password_hints_allowed": false,
  "password_iterations": 600000,
  "push_enabled": true,
  "push_installation_id": "***",
  "push_installation_key": "***",
  "push_relay_uri": "https://push.bitwarden.com",
  "reload_templates": false,
  "require_device_email": false,
  "rsa_key_filename": "data/rsa_key",
  "send_purge_schedule": "0 5 * * * *",
  "sendmail_command": null,
  "sends_allowed": true,
  "sends_folder": "data/sends",
  "show_password_hint": false,
  "signups_allowed": false,
  "signups_domains_whitelist": "******,******************",
  "signups_verify": true,
  "signups_verify_resend_limit": 3,
  "signups_verify_resend_time": 3600,
  "smtp_accept_invalid_certs": false,
  "smtp_accept_invalid_hostnames": false,
  "smtp_auth_mechanism": null,
  "smtp_debug": false,
  "smtp_embed_images": true,
  "smtp_explicit_tls": null,
  "smtp_from": "****************",
  "smtp_from_name": "Vaultwarden",
  "smtp_host": "*****************",
  "smtp_password": null,
  "smtp_port": 25,
  "smtp_security": "off",
  "smtp_ssl": false,
  "smtp_timeout": 15,
  "smtp_username": null,
  "templates_folder": "data/templates",
  "tmp_folder": "data/tmp",
  "trash_auto_delete_days": 60,
  "trash_purge_schedule": "0 5 0 * * *",
  "use_sendmail": false,
  "use_syslog": false,
  "user_attachment_limit": 5121,
  "web_vault_enabled": true,
  "web_vault_folder": "web-vault/",
  "websocket_address": "0.0.0.0",
  "websocket_enabled": false,
  "websocket_port": 3012,
  "yubico_client_id": "61202",
  "yubico_secret_key": "***",
  "yubico_server": null
}
```
</details>

* Install method Docker

* Clients used: Directory Connector

* Reverse proxy and version: apache

* MySQL/MariaDB or PostgreSQL version: mariadbsu

* Other relevant details:

### Steps to reproduce

1. Create organisation and get API-Key
2. Configure Directory Connector to sync from AzureAD
3. After Users have been invited, remove one or more Users from the sync scope
4. Rerun sync

### Expected behaviour
After removing the user from the sync scope, the user should be removed from the organisation, when "Overwrite existing organization users based on current sync settings." is activated

### Actual behaviour
The User remains in the organisation and is not removed.

### Troubleshooting data
The "test now" also doesn´t list the user under "deleted Users". The logfiles also don´t show any errors

**Comment by BlackDex** [comment]
I have tested this using LDAP, and while it also doesn't show the users under the delete option for me, it does delete the actual user which is not in the list anymore. Just make sure have the correct flags set.

![image](https://github.com/dani-garcia/vaultwarden/assets/610450/95180dee-21d4-4b8e-a462-fdbb44f1816c)

Also, if it doesn't list them as deleted, you probably need to clear the previous sync cache, and if you do that, it will remove the user from the organization. It does not remove the user it self, so they can still login and access there personal vault items.

This is not something we can fix, as the actual workings on removing and adding users works just fine.
It's just that the Directory Connector does not make a call to the server to update the status until you clear the cache.

So, it's either a bug in the Directory Connector, or, as what i think a limitation of LDAP (In my case). Since i think AzureAD has some flags available to support disabeling a user, and if that flag is found, the Directory Connector will detect that also, and mark it as deletion.


It might be possible to add this feature to LDAP by configuring it in the right way. But it needs the `userAccountControl` attribute.
https://github.com/bitwarden/directory-connector/blob/a3c8629f6d4213b13b40bf0efcfcc33a76f2497c/src/services/ldap-directory.service.ts#L305-L318

And according to this page https://learn.microsoft.com/en-us/troubleshoot/windows-server/identity/useraccountcontrol-manipulate-account-properties#list-of-property-flags, it needs to be set to `2`

**Reply by bokkabonga** [reply]
I couldn´t test using LDAP so far, since it always throws erros at me. So my results are coming from syncing with AzureAD only.

I tried clearing the sync cache and have tested all combinations of settings. I´m also aware, that the users will only be removed from the organization not the server entirely.

Sadly this didnt change anything. The users are not deleted when removed from the groups that selected in the User Filter

**Reply by bokkabonga** [reply]
Got it to work with LDAP as well. Same behavior in that case. Removing a user from the group will not remove it from the org. Even with the options set as shown in your screenshot and after clearing cache.

**Reply by BlackDex** [reply]
If you only remove a user from a group, but that user is still fetched for that org, it will not be removed.
A user does not need to be part of a group before it can have access.

You need to fully remove that user from LDAP, it should not show in the user list during a test sync, if it still does, it will be kept there.

**Reply by bokkabonga** [reply]
I limited the LDAP search to the members of one LDAP-group via this User filter:
`(&(memberOf=cn=test-bitwarden,ou=Group,dc=company,dc=com))`
After removing the user from the group and clearing cache the user is not listed anymore during test sync. So that is working. But the user is not removed during the actual sync.

I don´t sync any LDAP-groups to bitwarden-groups, in case you understood it that way.

**Reply by BlackDex** [reply]
I quickly tested it, and it works just fine for me. If i add a `User Filter` `(&(memberOf=cn=managers,ou=groups,dc=vwrs,dc=intern))` it removes all users not a member if that group, and if remove that filter again and sync they will be added again, and i can do this swapping multiple times and they are removed and add each time over and over again.

**Reply by bokkabonga** [reply]
What happens if you first add all members of the group using your filter and then remove one member from the ldap group and sync again? Is that user removed from the org?

**Reply by BlackDex** [reply]
I did the following:
 - Changed the filter to use `cn=users,ou=groups,dc=vwrs,dc=intern`, that removed all the users of the previous filter and added users of the other filter.
 - I then removed a user from that group and pressed `Test Sync` and that showed that that user was removed.
 - Then i clicked on `Sync Now` and it removed the user not longer part of that group.
 - After that i added the user back again, synced and it added that user again.

So, it all seems to work just fine for me.

When using LDAP just make sure the
 - `This server uses Active Directory` is unchecked
 - `Remove disabled users during sync` is checked
 - `Overwrite existing organization users based on current sync settings.` is checked

**Reply by bokkabonga** [reply]
I tested excactly what you did:

- Set the filter to `(&(memberOf=cn=test-bitwarden,ou=Group,dc=company,dc=com))`. That invites all users in the group to the org.
- Changed the filter to `(&(memberOf=cn=test-bitwarden-2,ou=Group,dc=company,dc=com))`. and pressed `Test Sync`. That only showed the users of the second group as `Users`. No `Deleted Users` however. Pressing `Sync Now` will also lead to no changes in the org
- Removed a User from the group and pressed `Test Sync`. It now shows the `Users` without the removed user. But still now `Removed Users`. Hitting `Sync Now` still does nothing and no changes in the org

The settings have all been set just as you describe and I´ve also tested different combinations of the settings.
Edit: Just in case it matters. The users are only invited. I didn´t accept the invitation or anything.

**Comment by bokkabonga** [comment]
I know the thread is very old, but I found the problem :D

It happens when the organisation contains users, that have been added manually and not via the directory connector. Since the external ID is not set, the user will not be removed.

I´m not sure if the behavior on the official server is the same or if those users would be deleted. If the latter is the case, we should change that here as well.

**Reply by BlackDex** [reply]
There is a checkbox for that: `Overwrite existing organization users based on current sync settings.`
![image](https://github.com/dani-garcia/vaultwarden/assets/610450/6b83d3b3-e83c-4753-ae9c-fe43b2d2b3c4)

**Reply by bokkabonga** [reply]
Yeah i checked that. However it doesn´t work with users that have been manually added.

e.g. I manually added the user "test.person@domain.com". The user is however not part of my user filter in the directory connector. The users does not get deleted, even if I clear the cache and run a new sync.

When I add the user to the LDAP group, syncing will add him to the org. Whe i remove him from the LDAP group, the user will be removed accordingly.

**Reply by BlackDex** [reply]
If that user also is not part of the LDAP/AD, then that is normal behavior. The directory connector code doesn't delete users not part if it.

**Reply by bokkabonga** [reply]
> If that user also is not part of the LDAP/AD, then that is normal behavior. The directory connector code doesn't delete users not part if it.

He kinda is. At least there exists a User with that mail-address. But as long as i only add users via the directory connector, i will be fine :)
